mesonbuild/compilers/gcc_function_attributes.py

```diff
@@ -0,0 +1,122 @@
+# These functions are based on the following code:
+# https://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_gcc_func_attribute.m4,
+# which is licensed under the following terms:
+#
+#   Copyright (c) 2013 Gabriele Svelto <gabriele.svelto@gmail.com>
+#
+#   Copying and distribution of this file, with or without modification, are
+#   permitted in any medium without royalty provided the copyright notice
+#   and this notice are preserved.  This file is offered as-is, without any
+#   warranty.
+
+GCC_FUNC_ATTRIBUTES = {
+    'alias': '''
+        int foo(void) { return 0; }
+        int bar(void) __attribute__((alias("foo")));''',
+    'aligned':
+        'int foo(void) __attribute__((aligned(32)));',
+    'alloc_size':
+        'void *foo(int a) __attribute__((alloc_size(1)));',
+    'always_inline':
+        'inline __attribute__((always_inline)) int foo(void) { return 0; }',
+    'artificial':
+        'inline __attribute__((artificial)) int foo(void) { return 0; }',
+    'cold':
+        'int foo(void) __attribute__((cold));',
+    'const':
+        'int foo(void) __attribute__((const));',
+    'constructor':
+        'int foo(void) __attribute__((constructor));',
+    'constructor_priority':
+        'int foo( void ) __attribute__((__constructor__(65535/2)));',
+    'deprecated':
+        'int foo(void) __attribute__((deprecated("")));',
+    'destructor':
+        'int foo(void) __attribute__((destructor));',
+    'dllexport':
+        '__attribute__((dllexport)) int foo(void) { return 0; }',
+    'dllimport':
+        'int foo(void) __attribute__((dllimport));',
```

Comment:
GCC an Clang also support the MSVC-style `__declspec(dllimport)` and `__declspec(dllexport)` function attributes, and those are usually used instead of these. I think we should test those instead.

Reply:
Also, once we change this and fix it for MSVC, in the docs above we'll want to specify what syntax was tested, so people know that `dllimport` and `dllexport` function attributes also work for MSVC, and that they should use `__declspec(dllimport)`, etc.